core/src/replay_stage.rs

```diff
@@ -807,37 +823,120 @@ impl ReplayStage {
         (progress, heaviest_subtree_fork_choice)
     }
 
-    #[allow(dead_code)]
-    fn reset_duplicate_slots(
-        duplicate_slots_reset_receiver: &DuplicateSlotsResetReceiver,
+    fn dump_then_repair_correct_slots(
+        duplicate_slots_to_repair: &mut DuplicateSlotsToRepair,
         ancestors: &mut HashMap<Slot, HashSet<Slot>>,
         descendants: &mut HashMap<Slot, HashSet<Slot>>,
         progress: &mut ProgressMap,
         bank_forks: &RwLock<BankForks>,
+        blockstore: &Blockstore,
+        poh_bank_slot: Option<Slot>,
     ) {
-        for duplicate_slot in duplicate_slots_reset_receiver.try_iter() {
-            Self::purge_unconfirmed_duplicate_slot(
-                duplicate_slot,
-                ancestors,
-                descendants,
-                progress,
-                bank_forks,
-            );
+        if duplicate_slots_to_repair.is_empty() {
+            return;
         }
+
+        let root_bank = bank_forks.read().unwrap().root_bank();
+        // TODO: handle if alternate version of descendant also got confirmed after ancestor was
```

Comment:
Does this mean if we have a fork of multiple duplicate blocks that we've confirmed we will try to repair slots that were already dumped?

Reply:
I think what I was imagining was something like:

```
1 -> 2 -> 3  -> 4
         | ---3' -> 4' -> 5
```

I was imagining we got duplicate confirmed for `3`, then `4`. So `3` and `4` would both get added to the `duplicate_slots_to_repair` list.

Then here: https://github.com/solana-labs/solana/blob/b74ff2362aaf4fe4a14a0ce2cbc30f5c05dce33e/core/src/replay_stage.rs#L843, either of `3` or `4` could be purged first. This function should not panic under both orderings

Test added `test_dump_then_repair_correct_slots()` here: https://github.com/solana-labs/solana/pull/18160/commits/33fc32f09eb517d4028f75593d1b829c0cc064bc